File: dispatch.py
#!/usr/bin/env python

import http.server
import socketserver
import json
import traceback
import urllib.parse

DISPATCH_LISTEN = "0.0.0.0"
DISPATCH_PORT = 8099
DISPATCH_ROUTE_URL = "http://localhost:8099"

GAME_IP = "localhost"
GAME_PORT = 22102

GAME_RES_CUR_VERSION = 138541
DESIGN_DATA_CUR_VERSION = 138541

PLAYER_UID = 1

FEEDBACK_URL = "https://cdn.discordapp.com/attachments/441109559004889098/1043690740397641748/unknown.png"
NOTICES_URL = "https://cdn.discordapp.com/attachments/441109559004889098/1043690740397641748/unknown.png"
GUIDE_URL = "https://cdn.discordapp.com/attachments/441109559004889098/1043690740397641748/unknown.png"

GAME_RES_CUR_VERSION_BYTES = str(GAME_RES_CUR_VERSION).encode("utf8")
DESIGN_DATA_CUR_VERSION_BYTES = str(DESIGN_DATA_CUR_VERSION).encode("utf8")

class DispatchHandler(http.server.BaseHTTPRequestHandler):
	def version_string(self):
		return "broken pancake"

	def do_GET(self):
		url = urllib.parse.urlparse(self.path)
		query = urllib.parse.parse_qs(url.query)
		for k in query.keys():
			query[k] = query[k][-1]

		try:
			if url.path == "/query_region_list":
				self.do_query_region_list(url, query)
			elif url.path == "/sdk/login":
				self.do_sdk_login(url, query)
			elif url.path == "/sdk/token_login":
				self.do_sdk_token_login(url, query)
			elif url.path == "/query_cur_region":
				self.do_query_cur_region(url, query)
			elif url.path == "/game_res/cur_version.txt":
				self.send_response(200)
				self.end_headers()
				self.wfile.write(GAME_RES_CUR_VERSION_BYTES)
			elif url.path.startswith("/game_res/"):
				self.send_response(200)
				self.end_headers()
			elif url.path == "/design_data/cur_version.txt":
				self.send_response(200)
				self.end_headers()
				self.wfile.write(DESIGN_DATA_CUR_VERSION_BYTES)
			elif url.path == "/":
				self.send_response(200)
				self.end_headers()
				self.wfile.write(b"<head><style>@keyframes soggy{from{background-position:0 234px}to{background-position:375px 0}}body{background:url(\"/soggy_cat.png\") repeat;animation:soggy 2s linear 0s infinite}</style></head>\n")
			elif url.path == "/soggy_cat.png":
				self.send_response(200)
				self.send_header("Content-Type", "image/png")
				self.send_header("Content-Disposition", "inline")
				self.end_headers()
				with open("soggy_cat.png", "rb") as f:
					self.wfile.write(f.read())
			elif url.path == "/favicon.ico":
				self.send_response(404)
				self.wfile.write(b"404 Not Found\n")
				self.end_headers()
			else:
				self.send_response(400)
				self.end_headers()
				self.wfile.write(b"400 Bad Request\n")
		except Exception as ex:
			traceback.print_exception(ex)
			self.send_response(500)
			self.end_headers()
			self.wfile.write(b"500 Internal Server Error\n")

	def do_HEAD(self):
		url = urllib.parse.urlparse(self.path)
		query = urllib.parse.parse_qs(url.query)
		for k in query.keys():
			query[k] = query[k][-1]

		try:
			if url.path == "/game_res/cur_version.txt":
				self.send_response(200)
				self.send_header("Content-Length", str(len(GAME_RES_CUR_VERSION_BYTES)))
				self.end_headers()
			elif url.path.startswith("/game_res/"):
				self.send_response(200)
				self.send_header("Content-Length", "0")
				self.end_headers()
			elif url.path == "/design_data/cur_version.txt":
				self.send_response(200)
				self.send_header("Content-Length", str(len(DESIGN_DATA_CUR_VERSION_BYTES)))
				self.end_headers()
			else:
				self.send_response(400)
				self.wfile.write(b"400 Bad Request\n")
				self.end_headers()
		except Exception as ex:
			traceback.print_exc(ex)
			self.send_response(500)
			self.end_headers()
			self.wfile.write(b"500 Internal Server Error\n")

	def do_POST(self):
		self.send_response(501)
		self.end_headers()

	def do_query_region_list(self, url: urllib.parse.ParseResult, query: dict):
		self.send_response(200)
		self.send_header("Content-Type", "application/json")
		self.end_headers()

		REGION_TITLES = [
			"100% real cbt1 server",
			"Do you really think.",
			"That you will escape God's judgement for your sins."
		]

		query_region_list_rsp = {
			"retcode": 0,
			"clientCustomConfig": json.dumps({ "sdkenv": "2", "visitor": False, "devicelist": None }),
			# ^ these are ALL of keys it checks for in clientCustomConfig
			# "devicelist" seems completely unused?
			"regionList": list()
		}

		for i, region_title in enumerate(REGION_TITLES):
			query_region_list_rsp["regionList"].append({
				"dispatchUrl": f"{DISPATCH_ROUTE_URL:s}/query_cur_region",
				"name": f"region{i:d}",
				"title": region_title,
				"type": "DEV_PUBLIC"
			})

		self.wfile.write(json.dumps(query_region_list_rsp).encode("utf8"))
	
	LOGIN_JSON = {
		"retcode": 0,
		"data": {
			"uid": str(PLAYER_UID),
			"token": "token1",
			"email": "soggy:)",
		}
	}

	def do_sdk_login(self, url: urllib.parse.ParseResult, query: dict):
		self.send_response(200)
		self.send_header("Content-Type", "application/json")
		self.end_headers()
		self.wfile.write(json.dumps(self.LOGIN_JSON).encode("utf8"))

	def do_sdk_token_login(self, url: urllib.parse.ParseResult, query: dict):
		self.send_response(200)
		self.send_header("Content-Type", "application/json")
		self.end_headers()
		self.wfile.write(json.dumps(self.LOGIN_JSON).encode("utf8"))

	def do_query_cur_region(self, url: urllib.parse.ParseResult, query: dict):
		self.send_response(200)
		self.send_header("Content-Type", "application/json")
		self.end_headers()

		query_cur_region_rsp = {
			"retcode": 0,
			"regionInfo": {
				"gateserverIp": GAME_IP,
				"gateserverPort": GAME_PORT,
				"resourceUrl": f"{DISPATCH_ROUTE_URL:s}/game_res",
				"dataUrl": f"{DISPATCH_ROUTE_URL:s}/design_data",
				"feedbackUrl": FEEDBACK_URL,
				"bulletinUrl": NOTICES_URL,
				"handbookUrl": GUIDE_URL,
				"resVersion": GAME_RES_CUR_VERSION,
				"dataVersion": DESIGN_DATA_CUR_VERSION,
				# these are unused I think
				"clientCustomConfig": json.dumps({}),
				"regionCustomConfig": json.dumps({})
			}
		}

		self.wfile.write(json.dumps(query_cur_region_rsp).encode("utf8"))

httpd = None
try:
	socketserver.TCPServer.allow_reuse_address = True
	httpd = socketserver.TCPServer((DISPATCH_LISTEN, DISPATCH_PORT), DispatchHandler)
	print(f"dispatch listening on port {DISPATCH_PORT:d}")
	httpd.serve_forever()
except KeyboardInterrupt:
	print("\nKeyboardInterrupt")
finally:
	if httpd != None:
		httpd.server_close()

File: src/dispatch.cpp
// dispatch.cpp

// the dispatch server :)
// I don't really understand how the client's update checks work, but my dispatch implementation gets past them

#include "dispatch.hpp"

// C++
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <ctime>
#include <fstream>
#include <iterator>
#include <string>

// cpp-httplib
#include <httplib.h>

// rapidjson
#define RAPIDJSON_HAS_STDSTRING 1
#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>

// soggy
#include "log.hpp"
#include "configfile.hpp"
#include "binpack.hpp"

void do_query_region_list(const httplib::Request &req, httplib::Response &rsp) {
	rapidjson::StringBuffer buffer;
	rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);

	writer.StartObject();
	writer.Key("retcode"); writer.Int(0);
	writer.Key("clientCustomConfig"); writer.String("{\"sdkenv\":\"2\",\"visitor\":false,\"devicelist\":null}");
	// ^ these are ALL of keys it checks for in clientCustomConfig
	// "sdkenv" MUST be >= 2
	// "devicelist" seems completely unused?

	writer.Key("regionList");
	writer.StartArray();

	writer.StartObject();
	writer.Key("dispatchUrl"); writer.String(soggy_cfg.dispatch_route_url + "/query_cur_region");
	writer.Key("name"); writer.String("region0");
	writer.Key("title"); writer.String(soggy_cfg.dispatch_region_title);
	writer.Key("type"); writer.String("DEV_PUBLIC");
	writer.EndObject();

	writer.EndArray();
	writer.EndObject();

	rsp.set_content(buffer.GetString(), "application/json");
}

void do_common_login(const httplib::Request &req, httplib::Response &rsp) {
	rapidjson::StringBuffer buffer;
	rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);

	writer.StartObject();
	writer.Key("retcode"); writer.Int(0);
	writer.Key("data");
	writer.StartObject();
	writer.Key("uid"); writer.String(std::to_string(1));
	writer.Key("token"); writer.String("token1");
	writer.Key("email"); writer.String("soggy:)");
	writer.EndObject();
	writer.EndObject();

	rsp.set_content(buffer.GetString(), "application/json");
}

extern void lookup_account_by_name(std::string name);
extern void lookup_account_uid_and_token(int uid, std::string token);

void do_sdk_login(const httplib::Request &req, httplib::Response &rsp) {
	// ?account=lda:)&password=x&lang=en
	do_common_login(req, rsp);
}
void do_sdk_token_login(const httplib::Request &req, httplib::Response &rsp) {
	// ?uid=1&token=token1&lang=en
	do_common_login(req, rsp);
}

void do_query_cur_region(const httplib::Request &req, httplib::Response &rsp) {
	rapidjson::StringBuffer buffer;
	rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);

	writer.StartObject();
	writer.Key("retcode"); writer.Int(0);
	writer.Key("regionInfo");
	writer.StartObject();
	writer.Key("gateserverIp"); writer.String(soggy_cfg.dispatch_game_addr);
	writer.Key("gateserverPort"); writer.Int(soggy_cfg.dispatch_game_port);
	writer.Key("resourceUrl"); writer.String(soggy_cfg.dispatch_route_url + "/game_res");
	writer.Key("dataUrl"); writer.String(soggy_cfg.dispatch_route_url + "/design_data");
	writer.Key("feedbackUrl"); writer.String(soggy_cfg.dispatch_feedback_url);
	writer.Key("bulletinUrl"); writer.String(soggy_cfg.dispatch_notices_url);
	writer.Key("handbookUrl"); writer.String(soggy_cfg.dispatch_guide_url);
	writer.Key("resVersion"); writer.Int(soggy_cfg.common_game_res_version);
	writer.Key("dataVersion"); writer.Int(soggy_cfg.common_design_data_version);
	writer.Key("clientCustomConfig"); writer.String("{}");
	writer.Key("regionCustomConfig"); writer.String("{}");
	writer.EndObject();
	writer.EndObject();

	rsp.set_content(buffer.GetString(), "application/json");
}

httplib::Server server;

bool dispatch_queued_shutdown = false;

void http_static_handler(const httplib::Request &req, httplib::Response &rsp, const std::string &key) {
	try {
		const Entry *entry = binpack_get_file(key);
		rsp.set_content((char *)entry->data, entry->len, server.find_content_type(key));
	} catch (std::out_of_range &) {
		rsp.status = 404;
	}
}

void dispatch_server_main() {
	httplib::Headers default_headers;
	default_headers.insert({ "Server", "broken pancake (real)" });
	server.set_default_headers(default_headers);

	server.Get("/static/(.*)", [](const httplib::Request &req, httplib::Response &rsp) {
		http_static_handler(req, rsp, req.matches[1].str());
	});

	server.Get("/query_region_list", do_query_region_list);
	server.Get("/sdk/login", do_sdk_login);
	server.Get("/sdk/token_login", do_sdk_token_login);
	server.Get("/query_cur_region", do_query_cur_region);

	server.Get("/game_res/cur_version.txt", [](const httplib::Request &req, httplib::Response &rsp) {
		rsp.set_content(std::to_string(soggy_cfg.common_game_res_version), "text/plain");
	});

	server.Get("/game_res/.*", [](const httplib::Request &req, httplib::Response &rsp) {});

	server.Get("/design_data/cur_version.txt", [](const httplib::Request &req, httplib::Response &rsp) {
		rsp.set_content(std::to_string(soggy_cfg.common_design_data_version), "text/plain");
	});

	server.Get("/", [](const httplib::Request &req, httplib::Response &rsp) {
		http_static_handler(req, rsp, "root.html");
	});

	soggy_log("dispatch listening on port %d", soggy_cfg.dispatch_bind_port);

	while (!dispatch_queued_shutdown) {
		server.listen(soggy_cfg.dispatch_bind_addr, soggy_cfg.dispatch_bind_port); // forever
	}

	server.stop();
}

void interrupt_dispatch_server() {
	server.stop();
	dispatch_queued_shutdown = true;
}
